### PR TITLE
Create and use  `NetworkKind`

### PR DESCRIPTION
**Pull request by tcharding**
OMG this is amazing, check this out.

On top of:
- #1832
- #1950
- #2200

Add a `NetworkKind` type and start using it:
- In BIP-32
- In `Address`
- In `WifKey`

This all came out of a leisurely Friday afternoon session investigating #2169. 

@dpc you are going to wet your pants.

### CODE REVIEW

File: bitcoin/src/address/mod.rs [hunk 1]
```diff
@@ -432,32 +434,30 @@ impl Address {
...
+        Self(AddressInner::P2pkh { hash, prefix: network.into() }, PhantomData)
     }
 
     /// Creates a pay to script hash P2SH address from a script.
     ///
     /// This address type was introduced with BIP16 and is the popular type to implement multi-sig
     /// these days.
     #[inline]
-    pub fn p2sh(script: &Script, network: Network) -> Result<Address, Error> {
+    pub fn p2sh(script: &Script, network: NetworkKind) -> Result<Address, Error> {
```
Comment (Kixunil) [inline bitcoin/src/address/mod.rs:451]: `impl Into<NetworkKind>` would be nicer I think.

**Review by Kixunil — COMMENTED** [review]
Hmm, this does look quite interesting. I will think about it more. But if we're going to do that I would suggest `Main` instead of `Real` since I think readers will have easier time understanding it..

**Comment by dpc** [comment]
> @dpc you are going to wet your pants.

This does not seem like an RFC where Bitcoin community obsoletes and recommend not using anymore address network-specific constants other than mainnet and testnet (or any arbitrary non-mainnet). :smile: 

Look. `Real` confused me for a moment as well.

**Comment by tcharding** [comment]
Superseded by #2232 but excluding the WIF key stuff which is a separate concern.

**Pull request closed without merging**